wal/wal.go

```diff
@@ -480,6 +497,90 @@ func (w *WAL) ReadAll() (metadata []byte, state raftpb.HardState, ents []raftpb.
 	return metadata, state, ents, err
 }
 
+// Verify reads through the records of the WAL and verifies that it is not corrupt.
+// Unlike ReadAll, it does not return the entries in the WAL.
+// It creates a new decoder to read through the records of the WAL.
+// It cannot be called after a ReadAll call on the WAL.
+// If it cannot read out the expected snap, it will return ErrSnapshotNotFound.
+// If the loaded snap doesn't match with the expected one, it will
+// return error ErrSnapshotMismatch.
+func (w *WAL) Verify() error {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	var metadata []byte
+	var err error
+	var match bool
+
+	rec := &walpb.Record{}
+
+	// create a new decoder
+	names, nameIndex, err := selectWALFiles(w.lg, w.dir, w.start)
```

Comment:
the method is attached to an open WAL. why do we need to select WAL files again? If we do need to do that, probably the Verify func should be a function rather than a method on the WAL struct?

Reply:
In that case, `Verify` function will not be able to use the mutex attached with the WAL struct. Simultaneous calls to `Verify` and/or `ReadAll` would cause data races, especially when WAL is opened in write mode.
Other than this, it makes sense to write `Verify` as a separate function. WDYT?

Reply:
> Other than this, it makes sense to write Verify as a separate function. WDYT?

Probably just document that it is not safe to double open WAL files or open WAL file and Verify it at the same time. Today the WAL might be corrupted if two etcd processes open it at the same time. We should flock it to prevent corruption due to race.

Reply:
I have re-written `Verify` as a function now, so it is no longer associated with an open WAL. It internally opens the WAL in read mode to create the decoder and then closes those readers.

I also observed that opening WAL twice, either both in read mode or one in read mode, other in write mode, wouldn't actually cause any race. Opening WAL twice in write mode was fatal. As `Verify` opens WAL in read mode, it doesn't actually conflict with an already open WAL, but nonetheless I have documented that it is not recommended to run `Verify` when the WAL has been already opened in write mode.